### PR TITLE
Add files via upload - HDS/PS Interface

### PR DESCRIPTION
**Pull request by ZacSpec**
Interface for Pyscannerbit is working. It still has one problem, which I'll submit an issue for.

### CODE REVIEW

**Comment by bstienen** [comment]
Is the issue you are having issue #12?

**Comment by ZacSpec** [comment]
Yes, issue #12.

File: examples/pyscannerbit_hds_interface.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def __call__(self, function):
+        # Get ranges of the test function. The 0.001 moves the minima 0.001 up
+        # and the maxima 0.001 down, in order to make use the sampling is not
+        # by accident moving outside of the test function range.
+        ranges = function.get_ranges(0.01)
+        ranges = np.array(ranges).tolist()
+        
+        simple = function.get_simple_interface()
+        # Can this be done without specifiying dimensionality? HDS and PS are looking for max/min respectively.
+        neg_f = lambda x,y: -simple(x,y)
```
Comment (ZacSpec) [inline examples/pyscannerbit_hds_interface.py]: Solution to issue #12: Replace with

if dimensions == 1:
            neg_f = lambda a: -simple(a)
        elif dimensions == 2:
            neg_f = lambda a,b: -simple(a,b)
        elif dimensions == 3:
            neg_f = lambda a,b,c: -simple(a,b,c)
        elif dimensions == 4:
etc.

Comment (bstienen) [reply examples/pyscannerbit_hds_interface.py]: See #12

**Comment by ZacSpec** [comment]
I've made the requested changes, and found a way to manually define function argument names so pyscannerbit knows the dimensionality of the simple function wrapper.

**Comment by bstienen** [comment]
Nice! Thanks for the edit. I have two last points that I hope you could address.

- The first one is that I tried to run the example, but I apparently don't have pyscannerbit installed. When trying to install it from the source on github (bjfar/pyscannerbit) I get error messages

    `The source directory /tmp/pip-req-build-asmh1wz8/yaml-cpp does not contain a CMakeLists.txt file.`

    Now could just be a mistake that I made, but just to make the experience for external users as easy as possible, could you include a "how to run" at the top of the file? Having an example is nice, but if it is hard to get running, this might stop people from using your example.

- Second change is the request to add the class to the appropriate submodule posterior or optimiser and import it from there. This cleans up the code a bit and makes it easier for other people to use your interface.

**Comment by ZacSpec** [comment]
Made the requested changed. I'm not familiar with how Travis CI works, so I'm not sure why its failing. Also, there is an MPI issue with the twalk scanner, so it won't run. The other scanners are working fine.

**Comment by ZacSpec** [comment]
Made the Flake8 formatting changes, but now the tests aren't working, as Pyscannerbit needs to be installed for the HdsPsInterface class in the optimisation module to be tested. Tried adding Pyscannerbit to the installed packages in setup.py, but I'm running into issues with the path to the fortran compiler. Also not sure if we want to mandate installing Pyscannerbit.

**Comment by bstienen** [comment]
I agree that we might not want to mandate Pyscannerbit installed. My suggestion would be to exclude this script from the unit tests to solve the errors. If this is okay with you @williamjameshandley? In that case I will merge it into a new branch in which I will update the unit tests and then merge this branch into master.

**Comment by williamjameshandley** [comment]
I think you're right that we shouldn't expect users to have everything installed in order for it to work, (and should perhaps make pyscannerbit an optional dependency in the setup.py) but the test suite should definitely check everything that we're in theory supporting. It looks like the issue is that we need to specify some additional dependencies like gfortran in order to get this to work.

**Comment by williamjameshandley** [comment]
@ZacSpec , it looks like the line
```bash
Using Python interpreter version 3.6.7 for build.
Using Python libraries version 3.5.2 for Python backend support.
NOTE: You are using different Python versions for the interpreter and the libraries!
```

means that even though yaml is installed, it hasn't been installed for the earlier version. Is this a known issue for pyscannerbit?

**Comment by ZacSpec** [comment]
Yes, this is an issue both Martin and I had. We solved it by modifying the setup.py file to specify the python locations:

"Edit cmake_args in the setup.py file like this:

        cmake_args = ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + libout,
                      '-DCMAKE_VERBOSE_MAKEFILE:BOOL=OFF',
                      '-Wno-dev',
                      '-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=' + libout,
                      '-DSCANNERBIT_STANDALONE=True',
                      '-DCMAKE_INSTALL_RPATH={0}'.format(origin),
                      '-DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON',
                      '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON',
                      '-DCMAKE_INSTALL_PREFIX:PATH=' + libout,
                      '-DPYTHON_LIBRARY=/home/mwhi/anaconda3/envs/tf-cpu/lib/libpython3.6m.so',
                      '-DPYTHON_INCLUDE_DIR=/home/mwhi/anaconda3/envs/tf-cpu/include/python3.6m/',
                      '-DUSE_PYSB_YAML=1'
                 ]
"

**Comment by williamjameshandley** [comment]
That sounds like the right thing to do. Could you add that to the setup.py?

**Comment by ZacSpec** [comment]
We edited it in the pyscannerbit setup.py file, which doesn't exist here, as pyscannerbit is being built from PyPI. Also, this issue no longer seems to come up when installing from the git clone. Is there a way to get Travis to install Pyscannerbit that way, by running:

git clone https://github.com/bjfar/pyscannerbit.git
cd pyscannerbit
python setup.py sdist
pip install dist/pyscannerbit-0.0.31.tar.gz  

Note that this runs into the long term problem of the distribution number changing.

**Comment by williamjameshandley** [comment]
>"Edit cmake_args in the setup.py file like this:

What is `libout`? Is there a link to this fix?

**Comment by ZacSpec** [comment]
Libout is the directory. From the setup.py file:

        # Temporary directory where libraries and other built files should go
        # These will get automatically copied to the final install location
        print('extdir:', extdir)
        #quit()
        libout = str(extdir.parent.absolute()) + '/pyscannerbit'

There is no link.

I see you've been trying a lot of the same things I did when trying to get this to work. I also installed pybind11 separately, and had trouble with confusion between gfortran and ifort, which I also had installed, so I removed ifort from my .bashrc.

This is what got it working for Martin:

"1) git clone https://github.com/bjfar/pyscannerbit.git
2) In pyscannerbit directory:
git clone https://github.com/pybind/pybind11.git pybind11
git clone https://github.com/bjfar/yaml-cpp.git yaml-cpp

3) Edit cmake_args in the setup.py file like this:

        cmake_args = ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + libout,
                      '-DCMAKE_VERBOSE_MAKEFILE:BOOL=OFF',
                      '-Wno-dev',
                      '-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=' + libout,
                      '-DSCANNERBIT_STANDALONE=True',
                      '-DCMAKE_INSTALL_RPATH={0}'.format(origin),
                      '-DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON',
                      '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON',
                      '-DCMAKE_INSTALL_PREFIX:PATH=' + libout,
                      '-DPYTHON_LIBRARY=/home/mwhi/anaconda3/envs/tf-cpu/lib/libpython3.6m.so',
                      '-DPYTHON_INCLUDE_DIR=/home/mwhi/anaconda3/envs/tf-cpu/include/python3.6m/',
                      '-DUSE_PYSB_YAML=1'
                 ]
(Zach: note that I ran from within my tf-cpu anaconda environment, and have chosen the paths accordingly).

4) python setup.py sdist

5) pip install dist/pyscannerbit-0.0.24.tar.gz

6) python test_all_scanners.py"

**Comment by williamjameshandley** [comment]
Sometimes I hate travis... OK -- pyscannerbit now installs on travis with this setup. It turns out that all you need to do is ensure that pyyaml is installed before you try to install pyscannerbit. I then needed to install the relevant system dependencies for pyscannerbit, which was tedious to work out, but easy.

There is one [failing test](https://travis-ci.org/DarkMachines/high-dimensional-sampling/builds/618381286) associated with an erroneous `settings` argument. Hopefully @ZacSpec or @bstienen can address now that travis is working.

**Comment by ZacSpec** [comment]
Thanks @williamjameshandley . @bstienen , when I run the test ("python setup.py test") on my computer, it passes fine. And the scan class of pyscannerbit definitely wants a settings argument, so I don't know why its throwing this error.

**Comment by williamjameshandley** [comment]
This isn't just a Travis issue, as I the same error on my machine. What version of python are you using?

**Comment by ZacSpec** [comment]
I'm using Python 3.6

**Comment by williamjameshandley** [comment]
It doesn't seem that it's the python version that's causing this. Do you not get an error if you do a fresh install in a virtual environment?
```python
python -m venv venv
source venv/bin/activate
pip install pyyaml
python setup.py install
python setup.py test
```

**Comment by ZacSpec** [comment]
Its failing with the settings error now. Also building GAMBIT using two python versions.

   Using Python interpreter version 3.7 for build.
   Using Python libraries version 3.6.8 for Python backend support.
   NOTE: You are using different Python versions for the interpreter and the libraries!

**Comment by williamjameshandley** [comment]
>Also building GAMBIT using two python versions.
>
>Using Python interpreter version 3.7 for build.
>Using Python libraries version 3.6.8 for Python backend support.
>NOTE: You are using different Python versions for the interpreter and the libraries!

The python build warning looks like its harmless (as the messages suggest in the build script). 

>Its failing with the settings error now. 

That means that there's something installed in your system python which fixes the error. @bstienen can you think what might be causing [this error](https://travis-ci.org/DarkMachines/high-dimensional-sampling/jobs/618490703#L1858).

**Comment by ZacSpec** [comment]
Ok, Pyscannerbit code has changed so now I need to update the interface.

**Comment by bstienen** [comment]
> That means that there's something installed in your system python which fixes the error. @bstienen can you think what might be causing [this error](https://travis-ci.org/DarkMachines/high-dimensional-sampling/jobs/618490703#L1858).

> Ok, Pyscannerbit code has changed so now I need to update the interface.

@ZacSpec I suspect this was an answer to Will's question which you thereby answered for me :P? I agree that this is the interface with Pyscannerbit that does not have a `settings` argument for the [Scanner object](https://github.com/bjfar/pyscannerbit/blob/master/pyscannerbit/scan.py). Don't know how Pyscannerbit works though, so if you know how to fix it that would be awesome.

**Comment by ZacSpec** [comment]
Yeah @bstienen , the `settings` argument has changed to `scanner_options`, but there have been some other changes as well, and now its throwing a few new errors. I don't know how to fix it right now, but hopefully will soon.

**Comment by ZacSpec** [comment]
@bstienen, there have been some changes to pyscannerbit, which I'm working through. The first I need your help with. Can you add a version of the simple wrapper which takes an extra argument, called 'scan', and does nothing with it?

From pyscannerbit:
 ```
if list(par_names)[0] is not "scan":
                msg="\n\nError in signature of user-supplied log-likelihood function! The first argument must be named 'scan' and will be passed a reference to a global ScannerBit object that can be used to e.g. print extra information to the scan output file.\n"
                raise SanityCheckException(msg)
```
And an example:
```
def rastrigin(scan,x,y,z):
    X = [x,y,z]
    A = 10
    return - (A + sum([(x**2 - A * np.cos(2 * math.pi * x)) for x in X]))
```

**Comment by ZacSpec** [comment]
Not sure what's happening with Travis. Python v3.4 fails because PyScannerbit need >=3.5. 3.5 is timing out. And it looks like the scan is starting in both 3.6 and 3.7, but still reports a failure, possibly due to the different interpreter and library versions.

**Comment by bstienen** [comment]
I made a quick edit to your code to make it more maintainable. What are currently your problems and do you need help in any way, shape or form?

**Comment by ZacSpec** [comment]
Thanks for the edit @bstienen. The issue I'm having is that while @williamjameshandley got Pyscannerbit to install for Travis, we still can't call the scan function. Relevant error (from run #93.2) is
`ImportError: ('Python version mismatch: module was compiled for Python 3.5, but the interpreter version is incompatible: 3.6.7 (default, Nov 28 2018, 13:38:24) \n[GCC 5.4.0 20160609].',) (in subprocess)`

When I was trying to get Pyscannerbit to work earlier this year, I had this exact issue. Martin and I fixed it, after talking to Ben, the developer, by specifying the relevant python location for cmake. Searching online finds that cmake has issues picking the correct python versions.

The way to specify this for cmake is to clone the Pyscannerbit repository, and edit the setup.py file. There is a `cmake_args = [...]` in setup.py, and by adding 
```
'-DPYTHON_LIBRARY=/home/mwhi/anaconda3/envs/tf-cpu/lib/libpython3.6m.so',
'-DPYTHON_INCLUDE_DIR=/home/mwhi/anaconda3/envs/tf-cpu/include/python3.6m/',
```
to it (note that the paths are for Martin's machine), and then following the install instructions for pyscannerbit (`python setup.py sdist`, then `pip install dist/pyscannerbit-0.0.31.tar.gz`), the problem was solved.

This solution still works for me when I try it on a fresh Anaconda environment. It won't work when I use venv, as I can't find the correct paths to specify. 

The solution I tried today was to clone the entire Pyscannerbit repository to this one, and edit the paths manually (based on @williamjameshandley's previous commits and some guesswork). It didn't work, but I think this may be the only solution. 

Do you have any ideas? For example, what paths to include, and whether cloning the Pyscannerbit repo will work. Sorry for the long post. It's been a frustrating few days.

**Comment by williamjameshandley** [comment]
To what extent is this a problem with our interface, rather than one with pyscannerbit? Could any of the maintainers of that repository fix this issue upstream?

**Comment by ZacSpec** [comment]
Pyscannerbit is itself an interface (to GAMBIT), so I don't believe they can fix this. When I search for other instances of this issue with cmake, manually defining python paths is the recommended solution.

From our interfaces perspective, once pyscannerbit is installed correctly, the HdsPsInterface class works fine.

**Comment by ZacSpec** [comment]
@williamjameshandley If Travis can't properly install pyscannerbit, should we exempt it from the automated testing?

**Comment by williamjameshandley** [comment]
Probably, although this is far from ideal.

**Comment by ZacSpec** [comment]
I'm talking to the Pyscannerbit developer. In the meantime, I think we have to exempt it from the Travis testing. @bstienen, do you have a preferred way to do this?

**Comment by bstienen** [comment]
Sorry for my late reply.
I think the most straightforward way to do this is by including an if-statement in the test_posterior_implementations function in the test_implementations.py file that checks if the class `c` (in the for loop) equals your class. If so, `continue` to the next class. A comment refering to this discussion would be nice.

**Comment by bstienen** [comment]
Ah... my fault, I thought the unit test issue was easy to solve but requires a bit more than just that if-statement. If you give me write access to your repo I can fix it.

When testing the solution in my own repo (an trying to install the resulting package), I found out that I have trouble installing pyscannerbit (it complains about OpenMP and broken compilers, after having complained about half a douzen of other dependencies). If we require pyscannerbit to be installed (as it currently is defined in setup.py), the entire package will not be installed (at least on my machine). As we said earlier that we did not want to require pyscannerbit to be installed, is it an idea to remove it from the installation requirements?

**Comment by bstienen** [comment]
Okay, so I got rid of all error messages, except one. This was not a problem up to two days ago and  as far as I can tell and the appearance of this error seems to coincide with the new release of PyYAML. Two solutions:

- stop supporting Python 3.4;
- force a specific `pyyaml` version to be installed (one that does not have this error).

I prefer the first solution. What about you guys?
@Zac: I made some changes to the location of your implementation code and to the naming of the class. Additionally I added a check that makes sure that `pyscannerbit` was imported. Could you verify that the code still works?

**Comment by bstienen** [comment]
@ZacSpec If the code is still working, we can merge the pull request into the master branch 😃.

**Comment by ZacSpec** [comment]
Hey @bstienen , sorry about the delay. Yes, the code still works. Thanks for cleaning it up. I'll just add the extra info about how to install pyscannerbit to the example. Hopefully the Pyscannerbit developers get back with a better solution in the meantime.

**Comment by ZacSpec** [comment]
Done. Should be ready to go.

**Pull request merged**